sdks/java/core/src/main/java/org/apache/beam/sdk/options/ValueProvider.java

```diff
@@ -0,0 +1,226 @@
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.databind.deser.ContextualDeserializer;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+
+/**
+ * {@link ValueProvider} is an interface which abstracts the notion of
+ * fetching a value that may or may not be currently available.  This can be
+ * used to parameterize transforms that only read values in at runtime, for
+ * example.
+ */
+@JsonSerialize(using = ValueProvider.Serializer.class)
+@JsonDeserialize(using = ValueProvider.Deserializer.class)
+public interface ValueProvider<T> {
+  /**
+   * Return the value wrapped by this {@link ValueProvider}.
+   */
+  T get();
+
+  /**
+   * Whether the contents of this {@link ValueProvider} is available to validation
+   * routines that run at graph construction time.
+   */
+  boolean shouldValidate();
+
+  /**
+   * {@link StaticValueProvider} is an implementation of {@link ValueProvider} that
+   * allows for a static value to be provided.
+   */
+  @JsonSerialize(using = ValueProvider.Serializer.class)
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class StaticValueProvider<T> implements ValueProvider<T>, Serializable {
+    @Nullable
+    private final T value;
+
+    StaticValueProvider(@Nullable T value) {
+      this.value = value;
+    }
+
+    /**
+     * Creates a {@link StaticValueProvider} that wraps the provided value.
+     */
+    public static <T> StaticValueProvider<T> of(T value) {
+      StaticValueProvider<T> factory = new StaticValueProvider<>(value);
+      return factory;
+    }
+
+    @Override
+    public T get() {
+      return value;
+    }
+
+    @Override
+    public boolean shouldValidate() {
+      return true;
+    }
+  }
+
+  /**
+   * {@link RuntimeValueProvider} is an implementation of {@link ValueProvider} that
+   * allows for a value to be provided at execution time rather than at graph
+   * construction time.
+   *
+   * <p>To enforce this contract, if there is no default, users must only call
+   * {@link #get()} at execution time (after a call to {@link Pipeline#run}),
+   * which will provide the value of OPTIONS.
+   */
+  @JsonSerialize(using = ValueProvider.Serializer.class)
+  @JsonDeserialize(using = ValueProvider.Deserializer.class)
+  public static class RuntimeValueProvider<T> implements ValueProvider<T>, Serializable {
```

Comment:
if both implementations need to be `Serializable`, any reason not to just move it to the root VP interface?